test/contracts/OVM/chain/OVM_CanonicalTransactionChain.gas.spec.ts

```diff
@@ -0,0 +1,332 @@
+import { expect } from '../../../setup'
+
+/* External Imports */
+import { ethers } from 'hardhat'
+import { Signer, ContractFactory, Contract, BigNumber, providers } from 'ethers'
+import { TransactionResponse } from '@ethersproject/abstract-provider'
+import { smockit, MockContract } from '@eth-optimism/smock'
+import _ from 'lodash'
+
+/* Internal Imports */
+import {
+  makeAddressManager,
+  setProxyTarget,
+  FORCE_INCLUSION_PERIOD_SECONDS,
+  FORCE_INCLUSION_PERIOD_BLOCKS,
+  setEthTime,
+  NON_ZERO_ADDRESS,
+  remove0x,
+  getEthTime,
+  getNextBlockNumber,
+  increaseEthTime,
+  getBlockTime,
+  ZERO_ADDRESS,
+} from '../../../helpers'
+import { defaultAbiCoder, keccak256 } from 'ethers/lib/utils'
+
+const ELEMENT_TEST_SIZES = [1, 2, 4, 8, 16]
+const DECOMPRESSION_ADDRESS = '0x4200000000000000000000000000000000000008'
+const MAX_GAS_LIMIT = 8_000_000
+
+const getQueueLeafHash = (index: number): string => {
+  return keccak256(
+    defaultAbiCoder.encode(
+      ['bool', 'uint256', 'uint256', 'uint256', 'bytes'],
+      [false, index, 0, 0, '0x']
+    )
+  )
+}
+
+const getSequencerLeafHash = (
+  timestamp: number,
+  blockNumber: number,
+  data: string
+): string => {
+  return keccak256(
+    '0x01' +
+      remove0x(BigNumber.from(timestamp).toHexString()).padStart(64, '0') +
+      remove0x(BigNumber.from(blockNumber).toHexString()).padStart(64, '0') +
+      remove0x(data)
+  )
+}
+
+const getTransactionHash = (
+  sender: string,
+  target: string,
+  gasLimit: number,
+  data: string
+): string => {
+  return keccak256(encodeQueueTransaction(sender, target, gasLimit, data))
+}
+
+const encodeQueueTransaction = (
+  sender: string,
+  target: string,
+  gasLimit: number,
+  data: string
+): string => {
+  return defaultAbiCoder.encode(
+    ['address', 'address', 'uint256', 'bytes'],
+    [sender, target, gasLimit, data]
+  )
+}
+
+interface BatchContext {
+  numSequencedTransactions: number
+  numSubsequentQueueTransactions: number
+  timestamp: number
+  blockNumber: number
+}
+
+interface AppendSequencerBatchParams {
```

Comment:
We should consider combining this stuff into `core-utils` so that we have a single source of truth for these kinds of utilities

Reply:
strong +1 on this; also, the first three functions here appear to be shadowed when I pulled this locally.

Reply:
Uses `core-utils` stuff now.